File: fixtures/projects/e2e-app/src/workers/email.ts
import type { Logger } from "../utils/logger";
import { format } from "../utils/format";
import { enqueueNotification } from "./notification";

export type EmailPayload = {
	to: string;
	subject: string;
	body: string;
	priority: "low" | "normal" | "high";
};

export type EmailResult = {
	messageId: string;
	queuedAt: string;
	status: "sent" | "queued" | "failed";
};

let emailCounter = 0;

export async function sendEmail(payload: EmailPayload, logger: Logger): Promise<EmailResult> {
	const messageId = `email-${++emailCounter}`;
	logger.info("Sending email", { messageId, to: payload.to });

	if (payload.priority === "high") {
		await enqueueNotification({ channel: "email", recipient: payload.to, message: payload.subject }, logger);
	}

	return {
		messageId,
		queuedAt: format(new Date()),
		status: payload.to.includes("@") ? "sent" : "failed",
	};
}

export function main(): void {
	const logger = { info: () => {}, debug: () => {}, error: () => {} } as unknown as Logger;
	const payload: EmailPayload = { to: "admin@example.com", subject: "Worker started", body: "Email worker is running", priority: "normal" };
	sendEmail(payload, logger).catch(() => {});
}

File: fixtures/projects/e2e-python/src/workers/email.py
import json

from src.logging.logger import AppLogger
from src.workers.notification import enqueue_notification


def _serialize_payload(payload: dict[str, str]) -> str:
    return json.dumps(payload, sort_keys=True)


def send_email(payload: dict[str, str], logger: AppLogger) -> dict[str, str | bool]:
    logger.info("sending email", recipient=payload["to"])
    if payload.get("priority") == "high":
        enqueue_notification(
            {
                "channel": "email",
                "recipient": payload["to"],
                "message": payload["subject"],
            },
            logger,
        )
    return {
        "message_id": f"email-{len(_serialize_payload(payload))}",
        "status": "sent",
        "delivered": True,
    }


def main() -> None:
    logger = AppLogger(debug=False)
    send_email(
        {
            "to": "ops@example.com",
            "subject": "worker started",
            "body": "email worker ready",
            "priority": "normal",
        },
        logger,
    )
